sdk/identity/azure-identity/src/default_azure_credential.cpp

```diff
@@ -43,9 +43,12 @@ DefaultAzureCredential::DefaultAzureCredential(
   auto const managedIdentityCred = std::make_shared<ManagedIdentityCredential>(options);
   auto const azCliCred = std::make_shared<AzureCliCredential>(options);
 
+  // DefaultAzureCredential caches the selected credential, so that it can be reused on subsequent
+  // calls.
   m_impl = std::make_unique<_detail::ChainedTokenCredentialImpl>(
       GetCredentialName(),
-      ChainedTokenCredential::Sources{envCred, wiCred, managedIdentityCred, azCliCred});
+      ChainedTokenCredential::Sources{envCred, wiCred, managedIdentityCred, azCliCred},
```

Comment:
BTW, so how long does it take now to get the token from the credential that sits past the `ManagedIdentityCredential`?

I.e. if I am a customer with only theAzure CLI installed, and I run an SDK sample, how long do I wait, before the order was rearranged, after, and with this fix?

Reply:
I'd say, under 2 seconds is pretty good, under 5 is more or less ok, under 10 seconds is technically tolerable, more than that would be confusing for the people who run samples, also not the best experience, if you often recompile and re-run.

Reply:
There is a substantial win here (basically 6000x)!

> BTW, so how long does it take now to get the token from the credential that sits past the `ManagedIdentityCredential`?

Let's assume the chain of credentials is as follows, where only `envCred` is going to succeed (via `ClientSecretCredential`) and comes after managed identity credential, and just measuring the `dacCred.GetToken(...)` call:
> ChainedTokenCredential::Sources{wiCred, managedIdentityCred, envCred, azCliCred}

**Before (i.e. DAC without caching):** 130 ops/s
**After (i.e. DAC with caching):** 774,773 op/s

> I'd say, under 2 seconds is pretty good, under 5 is more or less ok, under 10 seconds is technically tolerable, more than that would be confusing for the people who run samples, also not the best experience, if you often recompile and re-run.

Just running the e2e azure-identity test locally, to see the first call experience, it is actually fairly quick with and without the caching (i.e. well under 1 second)! Note the timestamps. After this caching change it is almost instant now.

**Before:**
```text
[2023-11-10T03:22:43.3706160Z T: 31312] INFO  : Identity: DefaultAzureCredential: Created with the following credentials: WorkloadIdentityCredential, ManagedIdentityCredential, EnvironmentCredential, AzureCliCredential.
=== ATTEMPT 1 ===
[2023-11-10T03:22:43.8745744Z T: 31312] INFO  : Identity: DefaultAzureCredential: Successfully got token from EnvironmentCredential.
=== ATTEMPT 2 ===
[2023-11-10T03:22:43.8936397Z T: 31312] INFO  : Identity: DefaultAzureCredential: Successfully got token from EnvironmentCredential.
=== ATTEMPT3 ===
[2023-11-10T03:22:43.9112237Z T: 31312] INFO  : Identity: DefaultAzureCredential: Successfully got token from EnvironmentCredential.
OK
```

**After:**
```text
[2023-11-10T03:28:18.1340414Z T: 34272] INFO  : Identity: DefaultAzureCredential: Created with the following credentials: WorkloadIdentityCredential, ManagedIdentityCredential, EnvironmentCredential, AzureCliCredential.
=== ATTEMPT 1 ===
[2023-11-10T03:28:18.4760682Z T: 34272] INFO  : Identity: DefaultAzureCredential: Successfully got token from EnvironmentCredential. Reuse this credential for subsequent calls.
[2023-11-10T03:28:18.4766425Z T: 34272] DEBUG : Identity: DefaultAzureCredential: Save this credential at index 2 for subsequent calls.
=== ATTEMPT 2 ===
[2023-11-10T03:28:18.4772765Z T: 34272] INFO  : Identity: DefaultAzureCredential: Successfully got token from EnvironmentCredential. Reuse this credential for subsequent calls.
=== ATTEMPT3 ===
[2023-11-10T03:28:18.4779544Z T: 34272] INFO  : Identity: DefaultAzureCredential: Successfully got token from EnvironmentCredential. Reuse this credential for subsequent calls.
OK
```